frontend/public/components/routes/create-route.tsx

```diff
@@ -0,0 +1,302 @@
+          this.setState({
+            formData: {
+              ...this.state.formData,
+              error: errorMessage
+            }
+          });
+        }
+      });
+  }
+
+  handleChange: React.ReactEventHandler<HTMLInputElement> = event => {
+    this.setState({
+      formData: {
+        ...this.state.formData,
+        [event.currentTarget.name]: event.currentTarget.value
+      }
+    });
+  }
+
+  changeService = (serviceName: string) => {
+    const selectedService = _.find(this.state.services, { metadata: { name: serviceName } });
+    const portOptions = getPortOptions(selectedService);
+    const labels = selectedService.metadata.labels;
+    this.setState({
+      formData: {
+        ...this.state.formData,
+        service: serviceName,
+      },
+      labels,
+      portOptions,
+    });
+  }
+
+  changeTargetPort = (portName: string) => {
+    this.setState({
+      formData: {
+        ...this.state.formData,
+        targetPort: portName
+      }
+    });
+  }
+
+  toggleSection: React.ReactEventHandler<HTMLInputElement> = event => {
+    this.setState({
+      toggleData: {
+        ...this.state.toggleData,
+        [event.currentTarget.name]: event.currentTarget.checked
+      }
+    });
+  }
+
+  save = event => {
+    event.preventDefault();
+
+    const {
+      name,
+      hostname,
+      path,
+      service,
+      targetPort,
+    } = this.state.formData;
+    const {
+      namespace,
+      labels,
+    } = this.state;
+    const route = {
+      kind: 'Route',
+      apiVersion: 'route.openshift.io/v1',
+      metadata: {
+        name,
+        namespace,
+        labels,
+      },
+      spec: {
+        to: {
+          kind: 'Service',
+          name: service
+        },
+        host: hostname,
+        path,
+        port: {
+          targetPort,
+        }
+      }
+    };
+
+    this.setState({ inProgress: true });
+    k8sCreate(RouteModel, route)
+      .then(() => {
+        this.setState({inProgress: false});
+        history.push(resourcePathFromModel(RouteModel, name, namespace));
+      }, err => this.setState({
+        formData: {
+          ...this.state.formData,
+          error: err.message,
+        },
+        inProgress: false
+      }));
+  }
+
+  render() {
+    const title = 'Create Route';
+    const { loaded, services, portOptions } = this.state;
+    const serviceOptions = {};
+    _.each(_.sortBy(services, 'metadata.name'), ({ metadata: { name } }) => serviceOptions[name] = <ResourceName kind="Service" name={name} />);
+
+    return <React.Fragment>
+      <Helmet>
+        <title>{title}</title>
+      </Helmet>
+      <div className="co-m-nav-title">
+        <h1 className="co-m-pane__heading co-m-pane__heading--baseline">
+          <div className="co-m-pane__name">
+            {title}
+          </div>
+          <div className="co-m-pane__heading-link">
+            <Link to={`/k8s/ns/${this.state.namespace}/routes/new`}>From YAML</Link>
```

Comment:
Is this link to YAML editor something we want in your editors?

Reply:
Yeah, it's a pattern we've talked about where the create button by default takes you to the form, but on the form there is a link to create from YAML. I think it makes sense for secret forms, too (although that one has multiple forms, too)